File: packages/codex-session-relay/src/codex_session_relay/marker.py
"""The managed marker: create-once facts on a filesystem the relay database never sees.

A hook that only inspects registered relationships cannot see a missing registration. The marker
therefore exists BEFORE the relationship and is checkable without asking the relay anything, which
is the whole reason it is not a table: a question about a task nobody registered has to stay
answerable when there is no row to read.

Every fact is published once and never updated in place. Publication writes a sibling temp file,
fsyncs it, links it onto the target and unlinks the temp, so first-publication-wins is an operating
system fact rather than a convention a writer might forget. A writer that dies mid-write leaves an
orphan temp and no target, so readers see the fact as absent; two writers racing produce exactly one
EEXIST, so the loser learns synchronously that it lost instead of silently discarding the winner.

Identity here is deliberately strict. Missing, empty, blank and non-string values all name nothing,
and two records that name nothing are never a match. Stated that way because that is the form the
rule gets violated in: None == None is not evidence.

The layout and every rule in this module are fixed by skills/crw-run/references/hook-contract.md.
"""

import hashlib
import json
import os
import re
import uuid
from dataclasses import dataclass
from pathlib import Path

MARKER_ENV = "CODEX_SESSION_RELAY_MARKER_ROOT"
PRECEDENCE = ("flag", "env", "xdg", "home")
DIRECTORY_NAME = "codex-session-marker"

PUBLISHED = "published"
EXISTS = "exists"

# The three single facts, and the three directories of numbered facts. Claims and dispositions are
# keyed by the writer's own session instead, because a writer publishes inside a directory it owns.
SINGLE_FACTS = {"intent": "intent.json", "bound": "bound.json", "relationship": "relationship.json"}
NUMBERED_FACTS = ("attempts", "conflicts", "resolutions")

CLAIM_FILE = "claim.json"

# An assignment id is the hex sha256 of a dispatch request id and nothing else. Validated rather
# than trusted, because the value reaches this module from a command line: a mistyped id would
# otherwise publish facts into a directory no reader can ever select, and a crafted one carrying
# path separators or a parent reference would publish them outside the workspace it names.
ASSIGNMENT_RE = re.compile(r"^[0-9a-f]{64}$")


def valid_assignment(assignment) -> bool:
    return bool(ASSIGNMENT_RE.match(str(assignment or "")))


@dataclass(frozen=True)
class MarkerSelection:
    """Which rule chose the marker root, and the exact value that won.

    Carried the way store.StateSelection carries its own reason: a participant that cannot say
    which rule applied to it cannot be compared with another participant.
    """

    path: Path
    source: str
    detail: str

    def to_record(self) -> dict:
        return {
            "path": str(self.path),
            "source": self.source,
            "detail": self.detail,
            "precedence": list(PRECEDENCE),
        }


def resolve_marker_root(explicit=None) -> MarkerSelection:
    """Highest precedence first: an explicit root, the environment, XDG, then the home default.

    Deliberately a DIFFERENT directory from the relay state directory. The contract gives the relay
    daemon no access to the marker at all, and a root nested inside the store's directory would make
    that separation a matter of good behaviour rather than of layout.
    """
    if explicit:
        return MarkerSelection(
            Path(explicit).expanduser().absolute(), "flag", "--marker-root " + str(explicit)
        )
    override = os.environ.get(MARKER_ENV)
    if override:
        return MarkerSelection(
            Path(override).expanduser().absolute(), "env", MARKER_ENV + "=" + override
        )
    xdg = os.environ.get("XDG_STATE_HOME")
    if xdg:
        base, source, detail = Path(xdg).expanduser(), "xdg", "XDG_STATE_HOME=" + xdg
    else:
        base, source = Path.home() / ".local" / "state", "home"
        detail = "default under " + str(Path.home() / ".local" / "state")
    return MarkerSelection((base / DIRECTORY_NAME).absolute(), source, detail)


def named(value) -> bool:
    """Whether a record actually names an identity.

    Missing, empty, blank and non-string values all name nothing, and collapsing them to one answer
    is the point: two records that name nothing must never compare equal, which is exactly what
    None == None quietly does.
    """
    return isinstance(value, str) and bool(value.strip())


def same_identity(left, right) -> bool:
    """Do two records name the same identity? Unnamed on either side is never a match."""
    return named(left) and named(right) and left == right


def workspace_key(workspace) -> str:
    """One directory per workspace, keyed on the resolved path.

    realpath rather than the spelling handed to the hook, so a symlinked or relative cwd reaches the
    same assignment the coordinator declared against rather than a fresh empty one.
    """
    return hashlib.sha256(str(Path(workspace).expanduser().resolve()).encode("utf-8")).hexdigest()


def assignment_id(dispatch_request_id: str) -> str:
    """The assignment directory name: the hash of the dispatch request id, never the id itself.

    Storing it in the clear would make correlation meaningless, because any session able to read the
    directory could then present it. The child holds the preimage from its own dispatch.
    """
    return hashlib.sha256(str(dispatch_request_id).encode("utf-8")).hexdigest()


def workspace_dir(root, workspace) -> Path:
    return Path(root) / workspace_key(workspace)


def assignment_dir(root, workspace, assignment) -> Path:
    """A workspace path outlives the assignment that used it, so the assignment owns a level.

    Named by the workspace alone, a reused worktree would resolve to the previous assignment: every
    fact there is create-once, so the new coordinator's writes would all fail EEXIST while the new
    child read the previous session's bind and released. The undeclared turn held on a fresh
    workspace would go unheld on a reused one.
    """
    return workspace_dir(root, workspace) / _checked_assignment(assignment)


def _checked_assignment(assignment) -> str:
    if not valid_assignment(assignment):
        raise ValueError(
            "an assignment id is the hex sha256 of a dispatch request id, not " + repr(assignment)
        )
    return str(assignment)


def fact_digest(payload: dict) -> str:
    """SHA-256 over the fact's JSON with factId removed, sorted keys, no whitespace, ASCII escaped.

    The exact spelling matters because an independent writer must reproduce it byte for byte.
    Pinned by the contract's reproducible vector, which the tests assert.
    """
    body = {key: value for key, value in payload.items() if key != "factId"}
    return hashlib.sha256(_canonical(body).encode("utf-8")).hexdigest()


def _canonical(payload) -> str:
    return json.dumps(payload, sort_keys=True, separators=(",", ":"))


def _fsync_directory(directory) -> None:
    """A linked name is not durable until its directory is.

    Best effort on purpose: some filesystems refuse to open a directory for fsync, and failing the
    publication over that would turn a durability improvement into an availability regression. The
    link itself already decided who won.
    """
    try:
        descriptor = os.open(directory, os.O_RDONLY)
    except OSError:
        return
    try:
        os.fsync(descriptor)
    except OSError:
        pass
    finally:
        os.close(descriptor)


def publish(target, payload: dict) -> str:
    """Create-once publication. Returns 'published' when this writer won, 'exists' when it lost.

    The temp sibling is what forces the writable unit to be the DIRECTORY rather than the single
    name, which is why a claim lives at claims/<session>/claim.json and not claims/<session>.json.
    Its name is dot-prefixed so a reader walking a fact directory never mistakes a half-written temp
    for a fact.

    Losing is returned rather than raised because it is an ordinary outcome: a replayed bind of the
    same identity is a no-op, and only the caller knows whether the value it lost to agrees.
    """
    target = Path(target)
    directory = target.parent
    directory.mkdir(mode=0o700, parents=True, exist_ok=True)
    temp = directory / (
        "." + target.name + ".tmp." + str(os.getpid()) + "." + uuid.uuid4().hex[:12]
    )
    descriptor = os.open(temp, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o600)
    try:
        os.write(descriptor, _canonical(payload).encode("utf-8"))
        os.fsync(descriptor)
    finally:
        os.close(descriptor)
    try:
        os.link(temp, target)
        outcome = PUBLISHED
    except FileExistsError:
        outcome = EXISTS
    finally:
        try:
            os.unlink(temp)
        except OSError:
            pass
    _fsync_directory(directory)
    return outcome


def _read_fact(path):
    """Read one fact. Returns (value, readable). Unreadable is never reported as absent."""
    try:
        text = Path(path).read_text(encoding="utf-8")
    except OSError:
        return None, False
    try:
        return json.loads(text), True
    except ValueError:
        # Parsed nothing, so nothing is known about this fact. That is "I could not read it",
        # which must not be reported as "it is not there".
        return None, False


def _identified(value, fact_id):
    """Attach the factId the READER walked to.

    Never copied out of a body: a factId a writer chose would prove nothing, and every coverage
    decision in the contract is keyed on this value.
    """
    if isinstance(value, dict):
        record = dict(value)
        record["factId"] = fact_id
        return record
    # Left exactly as it is. A fact that is not a record has to reach the malformed check as the
    # wrong shape; coercing it here would hide the one thing the reader must report.
    return value


def read_assignment(directory):
    """Every published fact in one assignment, plus the labels of anything that could not be read.

    A reader arriving mid-race sees a subset of files, which is always a valid earlier state rather
    than a corrupt one, so absence is never an error here. Unreadability is, and it is returned
    separately because "I could not look" and "there is nothing there" are different answers.
    """
    directory = Path(directory)
    marker, unreadable = {}, []

    for key, name in SINGLE_FACTS.items():
        path = directory / name
        if not path.exists():
            continue
        value, readable = _read_fact(path)
        if not readable:
            unreadable.append(key)
            continue
        marker[key] = _identified(value, key)

    for key in NUMBERED_FACTS:
        sub = directory / key
        if not sub.is_dir():
            continue
        items = []
        for entry in sorted(sub.glob("*.json")):
            if entry.name.startswith("."):
                continue
            value, readable = _read_fact(entry)
            if not readable:
                unreadable.append(key + "/" + entry.stem)
                continue
            items.append(_identified(value, key + "/" + entry.stem))
        marker[key] = items

    claims_dir = directory / "claims"
    if claims_dir.is_dir():
        claims = []
        for session_dir in sorted(p for p in claims_dir.iterdir() if p.is_dir()):
            path = session_dir / CLAIM_FILE
            if not path.exists():
                continue
            value, readable = _read_fact(path)
            fact_id = "claims/" + session_dir.name + "/" + CLAIM_FILE
            if not readable:
                unreadable.append(fact_id)
                continue
            claims.append(_identified(value, fact_id))
        marker["claims"] = claims

    return marker, unreadable


def read_disposition(directory, session_id, turn_id):
    """The disposition this session recorded for this turn, if it published one.

    Read at the path the reader's own Stop identity derives rather than by searching for a body that
    matches, so a disposition published for another turn can never answer for this one.
    """
    if not (named(session_id) and named(turn_id)):
        return None, True
    path = Path(directory) / "dispositions" / session_id / (turn_id + ".json")
    if not path.exists():
        return None, True
    value, readable = _read_fact(path)
    if not readable:
        return None, False
    return _identified(value, "dispositions/" + session_id + "/" + turn_id), True


def list_assignments(root, workspace):
    """Every assignment directory declared for this workspace, oldest name first.

    Sorted by name so that every reader of the same listing walks it in the same order, which is
    what makes the contract's tie-break on assignment id reproducible.
    """
    directory = workspace_dir(root, workspace)
    try:
        return sorted(p for p in directory.iterdir() if p.is_dir())
    except OSError:
        return []

File: internal/relay/delivery/marker.go
package delivery

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"syscall"
	"unicode/utf8"

	"github.com/thisisjun786/codex-relay-workflow/internal/relay/store"
)

// The managed marker (marker.py): create-once facts on a filesystem the relay database never
// sees. Every fact is published once, by writing a sibling temp, fsyncing it and linking it onto
// the target, so first-publication-wins is an operating-system fact; the layout and every rule
// here are fixed by skills/crw-run/references/hook-contract.md.

// Marker words and names (marker.py).
const (
	MarkerEnv           = "CODEX_SESSION_RELAY_MARKER_ROOT"
	markerDirectoryName = "codex-session-marker"
	Published           = "published"
	Exists              = "exists"
	claimFile           = "claim.json"
)

// MarkerPrecedence is PRECEDENCE: the order the marker root rules are asked in.
var MarkerPrecedence = []string{"flag", "env", "xdg", "home"}

var singleFacts = []struct{ key, name string }{{"intent", "intent.json"}, {"bound", "bound.json"}, {"relationship", "relationship.json"}}
var numberedFacts = []string{"attempts", "conflicts", "resolutions"}

var assignmentPattern = regexp.MustCompile(`^[0-9a-f]{64}$`)

// ValidAssignment is valid_assignment: the hex sha256 of a dispatch request id.
func ValidAssignment(assignment any) bool {
	text, _ := assignment.(string)
	return assignmentPattern.MatchString(text)
}

// Named is named: only a string with a non-blank character (str.strip) names an identity.
func Named(value any) bool {
	text, ok := value.(string)
	return ok && strings.TrimFunc(text, isPySpace) != ""
}

// isPySpace is str.isspace for one character, the set str.strip removes.
func isPySpace(r rune) bool {
	switch r {
	case ' ', '\t', '\n', '\v', '\f', '\r', 0x1c, 0x1d, 0x1e, 0x1f, 0x85, 0xa0, 0x1680, 0x2028, 0x2029, 0x202f, 0x205f, 0x3000:
		return true
	}
	return r >= 0x2000 && r <= 0x200a
}

// SameIdentity is same_identity: unnamed on either side is never a match.
func SameIdentity(left, right any) bool {
	return Named(left) && Named(right) && left == right
}

// ValidSegment is valid_segment: an identity that may be used as a directory name.
func ValidSegment(value any) bool {
	if !Named(value) {
		return false
	}
	text := value.(string)
	if text == "." || text == ".." {
		return false
	}
	return !strings.ContainsAny(text, "/\\\x00")
}

// MarkerSelection is marker.MarkerSelection: which rule chose the root, and the value that won.
type MarkerSelection struct{ Path, Source, Detail string }

// Record is MarkerSelection.to_record.
func (m MarkerSelection) Record() Obj {
	precedence := make([]any, len(MarkerPrecedence))
	for i, p := range MarkerPrecedence {
		precedence[i] = p
	}
	return Obj{{Key: "path", Value: m.Path}, {Key: "source", Value: m.Source}, {Key: "detail", Value: m.Detail}, {Key: "precedence", Value: precedence}}
}

// ResolveMarkerRoot is resolve_marker_root: flag, then the environment, then XDG, then home.
// Deliberately a different directory from the relay state directory.
func ResolveMarkerRoot(explicit string) (MarkerSelection, error) {
	if explicit != "" {
		path, err := absoluteUser(explicit)
		return MarkerSelection{path, "flag", "--marker-root " + explicit}, err
	}
	if override := os.Getenv(MarkerEnv); override != "" {
		path, err := absoluteUser(override)
		return MarkerSelection{path, "env", MarkerEnv + "=" + override}, err
	}
	var base, source, detail string
	if xdg := os.Getenv("XDG_STATE_HOME"); xdg != "" {
		expanded, err := store.ExpandUser(xdg)
		if err != nil {
			return MarkerSelection{}, err
		}
		base, source, detail = expanded, "xdg", "XDG_STATE_HOME="+xdg
	} else {
		home, err := os.UserHomeDir()
		if err != nil {
			return MarkerSelection{}, err
		}
		base, source = filepath.Join(home, ".local", "state"), "home"
		detail = "default under " + base
	}
	path, err := absolutePath(base + "/" + markerDirectoryName)
	return MarkerSelection{path, source, detail}, err
}

// absoluteUser is Path(value).expanduser().absolute(): no symlink is resolved and no ".." folded.
func absoluteUser(value string) (string, error) {
	expanded, err := store.ExpandUser(value)
	if err != nil {
		return "", err
	}
	return absolutePath(expanded)
}

// absolutePath is Path.absolute(): the pathlib spelling, prefixed by the cwd when relative.
func absolutePath(value string) (string, error) {
	if !strings.HasPrefix(value, "/") {
		cwd, err := os.Getwd()
		if err != nil {
			return "", err
		}
		value = cwd + "/" + value
	}
	parts := []string{}
	for _, part := range strings.Split(value, "/") {
		if part != "" && part != "." {
			parts = append(parts, part)
		}
	}
	return "/" + strings.Join(parts, "/"), nil
}

// resolved is Path(value).expanduser().resolve(): symlinks followed, a missing tail kept.
func resolved(value string) (string, error) {
	expanded, err := store.ExpandUser(value)
	if err != nil {
		return "", err
	}
	return store.ResolvePath(expanded)
}

func sha256Hex(text string) string {
	sum := sha256.Sum256([]byte(text))
	return hex.EncodeToString(sum[:])
}

// WorkspaceKey is workspace_key: sha256 of the resolved workspace path, so a symlinked or
// relative cwd reaches the same assignment the coordinator declared against.
func WorkspaceKey(workspace string) (string, error) {
	path, err := resolved(workspace)
	if err != nil {
		return "", err
	}
	return sha256Hex(path), nil
}

// AssignmentID is assignment_id: the hash of the dispatch request id, never the id itself.
func AssignmentID(dispatchRequestID string) string { return sha256Hex(dispatchRequestID) }

// WorkspaceDir is workspace_dir.
func WorkspaceDir(root, workspace string) (string, error) {
	key, err := WorkspaceKey(workspace)
	if err != nil {
		return "", err
	}
	return filepath.Join(root, key), nil
}

// AssignmentDir is assignment_dir: the assignment owns a level under its workspace. A value that
// is not an assignment id is ValueError, as _checked_assignment raises it.
func AssignmentDir(root, workspace string, assignment any) (string, error) {
	if !ValidAssignment(assignment) {
		return "", &hostError{"ValueError", "an assignment id is the hex sha256 of a dispatch request id, not " + pyReprValue(assignment)}
	}
	directory, err := WorkspaceDir(root, workspace)
	if err != nil {
		return "", err
	}
	return filepath.Join(directory, assignment.(string)), nil
}

// FactDigest is fact_digest: SHA-256 over the fact's canonical JSON with factId removed.
func FactDigest(payload Obj) string {
	body := Obj{}
	for _, f := range payload {
		if f.Key != "factId" {
			body = append(body, f)
		}
	}
	return sha256Hex(canonical(body))
}

// canonical is _canonical: json.dumps(sort_keys=True, separators=(",", ":")), ASCII escaped.
func canonical(value any) string {
	var b strings.Builder
	writeCanonical(&b, value)
	return b.String()
}

func writeCanonical(b *strings.Builder, value any) {
	switch v := value.(type) {
	case Obj:
		fields := append(Obj(nil), v...)
		sort.SliceStable(fields, func(i, j int) bool { return fields[i].Key < fields[j].Key })
		b.WriteByte('{')
		for i, f := range fields {
			if i > 0 {
				b.WriteByte(',')
			}
			writeString(b, f.Key)
			b.WriteByte(':')
			writeCanonical(b, f.Value)
		}
		b.WriteByte('}')
	case []any:
		b.WriteByte('[')
		for i, item := range v {
			if i > 0 {
				b.WriteByte(',')
			}
			writeCanonical(b, item)
		}
		b.WriteByte(']')
	default:
		writeJSON(b, v, true)
	}
}

// fsyncDirectory is _fsync_directory: best effort, the link already decided who won.
func fsyncDirectory(directory string) {
	handle, err := os.Open(directory)
	if err != nil {
		return
	}
	_ = handle.Sync()
	_ = handle.Close()
}

// Publish is marker.publish: create-once publication, "published" when this writer won and
// "exists" when it lost. root, when not empty, confines the write under it before and after the
// directory is made.
func Publish(target string, payload any, root string) (string, error) {
	directory := filepath.Dir(target)
	if root != "" {
		if _, err := confined(directory, root); err != nil {
			return "", err
		}
	}
	if err := os.MkdirAll(directory, 0o700); err != nil {
		return "", err
	}
	if root != "" {
		if _, err := confined(directory, root); err != nil {
			return "", err
		}
	}
	suffix := make([]byte, 6)
	if _, err := rand.Read(suffix); err != nil {
		return "", err
	}
	temp := filepath.Join(directory, "."+filepath.Base(target)+".tmp."+strconv.Itoa(os.Getpid())+"."+hex.EncodeToString(suffix))
	body := []byte(canonical(payload))
	handle, err := os.OpenFile(temp, os.O_CREATE|os.O_EXCL|os.O_WRONLY|syscall.O_NOFOLLOW, 0o600)
	if err != nil {
		return "", err
	}
	if _, err := handle.Write(body); err != nil {
		_ = handle.Close()
		_ = os.Remove(temp)
		return "", err
	}
	if err := handle.Sync(); err != nil {
		_ = handle.Close()
		_ = os.Remove(temp)
		return "", err
	}
	if err := handle.Close(); err != nil {
		_ = os.Remove(temp)
		return "", err
	}
	outcome := Published
	linkErr := os.Link(temp, target)
	_ = os.Remove(temp)
	switch {
	case errors.Is(linkErr, fs.ErrExist):
		outcome = Exists
	case linkErr != nil:
		return "", linkErr
	}
	fsyncDirectory(directory)
	return outcome, nil
}

// Fact read answers (_read_fact): present, absent, or unreadable.
const (
	factPresent = iota
	factAbsent
	factUnreadable
)

// readFact is _read_fact: "it is not there" and "I could not look" are different answers.
func readFact(path string) (any, int) {
	data, err := os.ReadFile(path)
	if errors.Is(err, fs.ErrNotExist) {
		return nil, factAbsent
	}
	if err != nil || !utf8.Valid(data) {
		return nil, factUnreadable
	}
	value, err := loads(string(data))
	if err != nil {
		return nil, factUnreadable
	}
	return value, factPresent
}

// listing is marker.listing: every entry of a directory, sorted, and whether it could be read.
// FileNotFound is the only absence. only is "" (any), "directories" or "entries".
func listing(directory, pattern, only string) ([]string, bool) {
	entries, err := os.ReadDir(directory)
	if errors.Is(err, fs.ErrNotExist) {
		return nil, true
	}
	if err != nil {
		return nil, false
	}
	var found []string
	for _, entry := range entries {
		if pattern != "" {
			if matched, _ := filepath.Match(pattern, entry.Name()); !matched {
				continue
			}
		}
		path := filepath.Join(directory, entry.Name())
		if only != "" {
			info, err := os.Stat(path)
			isDirectory := false
			if err == nil {
				isDirectory = info.IsDir()
			} else if !errors.Is(err, fs.ErrNotExist) {
				return nil, false
			}
			if (only == "directories") != isDirectory {
				continue
			}
		}
		found = append(found, path)
	}
	sort.Strings(found)
	return found, true
}

// confined is marker.confined: resolve a path and require it to stay under the root owning it.
func confined(path, root string) (string, error) {
	resolvedPath, err := store.ResolvePath(path)
	if err != nil {
		return "", err
	}
	base, err := store.ResolvePath(root)
	if err != nil {
		return "", err
	}
	if resolvedPath != base && !strings.HasPrefix(resolvedPath, strings.TrimSuffix(base, "/")+"/") {
		return "", &hostError{"ValueError", "refusing to write outside the marker root: " + resolvedPath + " is not under " + base}
	}
	return resolvedPath, nil
}

// identified is _identified: the factId the READER walked to, never one copied out of a body.
// A fact that is not a record reaches the reader as it is.
func identified(value any, factID string) any {
	record, ok := value.(Obj)
	if !ok {
		return value
	}
	out := append(Obj(nil), record...)
	return set(out, "factId", factID)
}

func stem(name string) string { return strings.TrimSuffix(name, filepath.Ext(name)) }

// ReadAssignment is read_assignment: every published fact, plus the labels of anything that could
// not be read. Nothing here fails; every outside-world step answers with a label instead.
func ReadAssignment(directory string) (Obj, []string) {
	marker, unreadable := Obj{}, []string{}
	for _, fact := range singleFacts {
		value, status := readFact(filepath.Join(directory, fact.name))
		switch status {
		case factAbsent:
			continue
		case factUnreadable:
			unreadable = append(unreadable, fact.key)
			continue
		}
		marker = append(marker, F{Key: fact.key, Value: identified(value, fact.key)})
	}
	for _, key := range numberedFacts {
		entries, readable := listing(filepath.Join(directory, key), "*.json", "")
		if !readable {
			unreadable = append(unreadable, key)
			continue
		}
		items := []any{}
		for _, entry := range entries {
			name := filepath.Base(entry)
			if strings.HasPrefix(name, ".") {
				continue
			}
			value, status := readFact(entry)
			switch status {
			case factAbsent:
				continue
			case factUnreadable:
				unreadable = append(unreadable, key+"/"+stem(name))
				continue
			}
			items = append(items, identified(value, key+"/"+stem(name)))
		}
		marker = append(marker, F{Key: key, Value: items})
	}
	sessions, readable := listing(filepath.Join(directory, "claims"), "", "directories")
	if !readable {
		return marker, append(unreadable, "claims")
	}
	claims := []any{}
	for _, session := range sessions {
		factID := "claims/" + filepath.Base(session) + "/" + claimFile
		value, status := readFact(filepath.Join(session, claimFile))
		switch status {
		case factAbsent:
			continue
		case factUnreadable:
			unreadable = append(unreadable, factID)
			continue
		}
		claims = append(claims, identified(value, factID))
	}
	return append(marker, F{Key: "claims", Value: claims}), unreadable
}

// ReadDisposition is read_disposition: the disposition this session recorded for this turn, read
// at the path the Stop identity derives. (nil, true) when absent or when the identity names no
// path; (nil, false) when it could not be read.
func ReadDisposition(directory string, sessionID, turnID any) (any, bool) {
	if !ValidSegment(sessionID) || !ValidSegment(turnID) {
		return nil, true
	}
	session, turn := sessionID.(string), turnID.(string)
	value, status := readFact(filepath.Join(directory, "dispositions", session, turn+".json"))
	switch status {
	case factAbsent:
		return nil, true
	case factUnreadable:
		return nil, false
	}
	return identified(value, "dispositions/"+session+"/"+turn), true
}

// ListAssignments is list_assignments: every assignment declared for this workspace, oldest name
// first, and whether the workspace directory could be read.
func ListAssignments(root, workspace string) ([]string, bool, error) {
	directory, err := WorkspaceDir(root, workspace)
	if err != nil {
		return nil, false, err
	}
	found, readable := listing(directory, "", "directories")
	return found, readable, nil
}

// markerFactList is a numbered or claims fact list as the reader returned it.
func markerFactList(marker Obj, key string) []any {
	v, _ := get(marker, key)
	list, _ := v.([]any)
	return list
}

func markerFact(marker Obj, key string) Obj {
	v, _ := get(marker, key)
	record, _ := v.(Obj)
	return record
}

func fieldOf(record Obj, key string) any {
	v, _ := get(record, key)
	return v
}
